lib/src/license_detection/license.dart

```diff
@@ -0,0 +1,283 @@
+/// to be in the form of [`<spdx-identifier>.txt`][1].
+/// The [directories] are not searched recursively.
+///
+/// Throws if any of the [directories] contains a file not meeting the above criteria's.
+///
+/// [1]: https://spdx.dev/ids/
+List<License> loadLicensesFromDirectories(Iterable<String> directories) {
+  var licenses = <License>[];
+
+  for (var dir in directories) {
+    Directory(dir).listSync(recursive: false).forEach((file) {
+      if (file.path.endsWith('.txt')) {
+        final license = licensesFromFile(file.path);
+        licenses.addAll(license);
+      } else {
+        throw FormatException(
+            'Invalid file type:\nExpected: "spdx-identifier" Actual: ${file.uri.pathSegments.last}');
+      }
+    });
+  }
+
+  return List.unmodifiable(licenses);
+}
+
+/// Returns a list of [License] from the given license file.
+///
+/// If license text contains the phrase `END OF TERMS AND CONDITIONS`
+/// which indicates the presence of optional text, two instances
+/// of [License] are returned.
+///
+/// Throws if file is not properly encoded or name of the file is an
+/// invalid [SPDX identifier][1].
+///
+/// [1]: https://github.com/spdx/license-list-XML/blob/master/DOCS/license-fields.md#explanation-of-spdx-license-list-fields
+@visibleForTesting
+List<License> licensesFromFile(String path) {
+  var licenses = <License>[];
+  final file = File(path);
+
+  final rawContent = file.readAsBytesSync();
+  final identifier = file.uri.pathSegments.last.split('.txt').first;
+
+  if (_invalidIdentifier.hasMatch(identifier)) {
+    throw ArgumentError(
+        'Invalid file name: expected: "path/to/file/<spdx-identifier>.txt" actual: $path');
+  }
+  var content = '';
+
+  try {
+    content = utf8.decode(rawContent);
+  } on FormatException {
+    throw ArgumentError('Invalid utf-8 encoding: $path');
+  }
+
+  licenses.add(License.parse(identifier, content));
+
+  // If a license contains a optional part create an additional license
+  // instance with the optional part of text removed to have
+  // better chances of matching.
+  if (content.contains(_endOfTerms)) {
+    final modifiedContent = content.split(_endOfTerms).first + _endOfTerms;
+    licenses.add(License.parse(identifier, modifiedContent));
+  }
+  return licenses;
+}
+
+/// Regex to match the all the text starting from `END OF TERMS AND CONDTIONS`.
+const _endOfTerms = 'END OF TERMS AND CONDITIONS';
+
+/// Generates crc-32 checksum for the given list of tokens
+/// by taking [granularity] token values at a time.
+@visibleForTesting
+List<NGram> generateChecksums(List<Token> tokens, int granularity) {
+  if (tokens.length < granularity) {
+    final text = tokens.join(' ');
+    return [NGram(text, crc32(utf8.encode(text)), 0, tokens.length - 1)];
+  }
+
+  var nGrams = <NGram>[];
+
+  for (var i = 0; i + granularity <= tokens.length; i++) {
+    var text = '';
+    tokens.skip(i).take(granularity).forEach((token) {
```

Comment:
This should be cheaper: `final text = tokens.skip(i).take(granularity).join(' ');` - although this will not add and extra whitespace at the end of the text. If you need the space, use `StringBuffer` instead:
https://api.dart.dev/stable/2.13.4/dart-core/StringBuffer-class.html

Reply:
To really speed this up we should not reference `tokens.value` at all. Rather make the crc32 function able to take a list of tokens and a normalized text, and hash the values directly from there. That would require the tokens to also have offsets into the normalized text...

Reply:
Still, while passing the list of token values won't we reference the `token.value`.

Reply:
What if we do this instead, 
`values = tokens.map((e) => e.value)`

Then we can access all the values by just indexing instead of traversing the list from start every time.

Reply:
> Still, while passing the list of token values won't we reference the token.value.

No - my idea would be to have something like:
```
class Token {
  final int startInNormalized;
  final int endInNormalized;
  [...]
}

int crc32(List<int> bytes, int start, int end) {
  var crc = 0xffffffff;;

  final length = end - start;
  for(var i = 0; i < length; i++) {
    crc = (crc >> 8) ^ _crc32Table[(crc & 255) ^ bytes[i]];
  }
}

/// Returns the crc32 checksum for each ngram of [granularity] adjacent tokens.
List<NGram> generateChecksums(List<Token> tokens, int granularity, List<int> normalizedBytes) {
  var nGrams = <NGram>[];

  for (var i = 0; i < tokens.length - granularity; i++) {
   final nGramStart =  tokens[i].start;
   final nGramEnd =  tokens[i + granularity].start;
    final crcValue = crc32(normalizedBytes, nGramStart, nGramEnd);
    nGrams.add(NGram(normalizedBytes, crcValue, nGramStart, nGramEnd));
  });
}
```

Reply:
Let's just run `dart run --pause-isolates-on-start --observe bin/detect_versions.dart` then open devtools in browser, record a CPU profile, run the program, stop recording CPU profile, and see if `generateChecksums` uses much CPU time...